mindmeld/resource_loader.py

```diff
@@ -537,131 +629,121 @@ def _update_query_file_dates(self, query_tree):
             # file existed before and now -> update
             old_file_info["modified"] = new_file_info["modified"]
 
-    def _build_word_freq_dict(self, **kwargs):  # pylint: disable=no-self-use
-        """Compiles unigram frequency dictionary of normalized query tokens
+    class WordFreqBuilder:
+        '''
+        Compiles unigram frequency dictionary of normalized query tokens
+        '''
+        def __init__(self, enable_stemming=False):
+            self.enable_stemming = enable_stemming
+            self.tokens = []
 
-        Args:
-            queries (list of Query): A list of all queries
-        """
-        enable_stemming = kwargs.get(ENABLE_STEMMING_ARGS)
-
-        # Unigram frequencies
-        tokens = []
-
-        for query in kwargs.get("queries"):
+        def add(self, query):
             for i in range(len(query.normalized_tokens)):
                 tok = query.normalized_tokens[i]
-                tokens.append(mask_numerics(tok))
-                if enable_stemming:
+                self.tokens.append(mask_numerics(tok))
+                if self.enable_stemming:
                     # We only add stemmed tokens that are not the same
                     # as the original token to reduce the impact on
                     # word frequencies
                     stemmed_tok = query.stemmed_tokens[i]
                     if stemmed_tok != tok:
-                        tokens.append(mask_numerics(stemmed_tok))
-
-        freq_dict = Counter(tokens)
-        return freq_dict
-
-    def _build_char_ngram_freq_dict(self, **kwargs):  # pylint: disable=no-self-use
-        """Compiles n-gram character frequency dictionary of normalized query tokens
-
-        Args:
-            queries (list of Query): A list of all queries
-        """
-        char_freq_dict = Counter()
-        for length, threshold in zip(kwargs.get("lengths"), kwargs.get("thresholds")):
-            if threshold > 0:
-                for q in kwargs.get("queries"):
+                        self.tokens.append(mask_numerics(stemmed_tok))
+
+        def get_resource(self):
+            return Counter(self.tokens)
+
+    class CharNgramFreqBuilder:
+        '''
+        Compiles n-gram character frequency dictionary of normalized query tokens
+        '''
+        def __init__(self, lengths, thresholds):
+            self.lengths = lengths
+            self.thresholds = thresholds
+            self.char_freq_dict = Counter()
+
+        def add(self, query):
+            for length, threshold in zip(self.lengths, self.thresholds):
+                if threshold > 0:
```

Comment:
@vembar It looks like these threshold are 0 for ANLP. Is that by design?

Reply:
Don't think we're using char_ngrams in ANLP right now, but setting threshold to 0 will essentially include every character ngram in the data as a feature [L501](https://github.com/cisco/mindmeld/blob/2ce240eb8d85cb979969e10f944a26c1f22679a9/mindmeld/models/query_features.py#L501). Trying to recall why we set the default threshold to 0 rather than 1. Seems unintuitive and also creates a feature set mismatch (We're still fine since sklearn vectorizer will ignore features not seen during training)? @serapio Do you remember?